src/msix/pack/AppxPackageWriter.cpp

```diff
@@ -64,7 +110,77 @@ namespace MSIX {
         APPX_PACKAGE_WRITER_PAYLOAD_STREAM_UTF8* payloadFiles, UINT64 memoryLimit) noexcept try
     {
         ThrowErrorIf(Error::InvalidState, m_state != WriterState::Open, "Invalid package writer state");
-        // TODO: implement
+        std::vector<std::unique_ptr<PayloadFile>> files;
+        // TODO: use memoryLimit for how many files are going to be added
+        for(UINT32 i = 0; i < fileCount; i++)
+        {
+            ComPtr<IStream> stream(payloadFiles[i].inputStream);
+            auto payloadFile = BuildPayloadFile(payloadFiles[i].fileName, stream, payloadFiles[i].contentType, payloadFiles[i].compressionOption);
+            files.push_back(std::move(payloadFile));
+        }
+        ProcessPayloadFiles(files);
+
         NOTIMPLEMENTED
     } CATCH_RETURN();
-}
\ No newline at end of file
+
+    std::unique_ptr<PayloadFile> AppxPackageWriter::BuildPayloadFile(const std::string& name, const ComPtr<IStream>& stream,

```

Comment:
This won't scale; you are holding the entire uncompressed package in memory.

Reply:
Potentially, depending on the call pattern.  But even if not, you are holding entire uncompressed files in memory.

---
In reply to: [278237881](https://github.com/Microsoft/msix-packaging/pull/95#discussion_r278237881) [](ancestors = 278237881)